### ISSUE TITLE
option to load specific layers - for ysort tilemap problem

### ISSUE THREAD

**Issue by blurymind** [notnullgames/tiled-kaboom#12]
How does one do y-sort with this extension? If I want the player/enemies/npc to be behind a tile, based on their y position?

**Comment by blurymind** [comment]
would be cool if we had a minimal demo project right here that uses this extension and solves this very issue!

**Comment by blurymind** [comment]
The current example demo project doesnt work - you get a black page with some 404 errors in console.

In the source code- say that player is a part of the tilemap data somehow 
https://github.com/notnullgames/tiled-kaboom/blob/main/index.html#L60
can the player be y sorted?

**Comment by konsumer** [comment]
> would be cool if we had a minimal demo project right here that uses this extension and solves this very issue!

Yeh! You should make one. I have a demo for the functionality of this library, but your specific usecase isn't something I'm personally trying to do with kaboom. Feel free to use this lib in your thing, and I can link to it in the README.

> The current example demo project doesnt work - you get a black page with some 404 errors in console.

Hmm. Seems like my demo broke due to changes on kaboom. I updated to use their current url, and it seems like it broke how levels work. I may not have time to fix it right now, but I will take a look.

I'm not sure I follow the other stuff you are saying. Sounds like things you could figure out if you make your own demo and play around with it a bit, but I'm not sure I fully understand what you are asking.

**Comment by blurymind** [comment]
thanks for having another look at the demo. It is still broken btw
![image](https://user-images.githubusercontent.com/6495061/137113632-5f60fe95-fc09-4212-92de-dce796d4e911.png)

**Comment by konsumer** [comment]
This really should be a separate issue, but I updated demo and published tiled-kaboom@0.0.12. They changed how `addLevel` works, but it was a simple enough fix.

**Comment by blurymind** [comment]
my idea for this is to update the demo itself, not the library :) 

For the library,it would be nice if the waterfall was animated, but there is another ticket for that

**Comment by blurymind** [comment]
Hey thanks for updating the demo. I will have a look in the morning. Yeah I
made the tiled plugin for gdevelop,  but am kind of hopeful other
developers will join in progressing it with time.

The kaboomjs extension here really interests me for the sole reason of
combining it with my tilemap editor at some point. I started writing a
tiled importer and exporter for it and needed something to motivate myself
this is going in the right direction.

Basically I want a kaboomjs dev environment in the browser with a tilemap
editor and this extension here is the missing link between the two. It is
getting closer

**Comment by konsumer** [comment]
> The kaboomjs extension here really interests me for the sole reason of combining it with my tilemap editor at some point.

I have no interest in implementing this, but you are free to use my code in your editor. Give credit, and use whatever you like.

> Basically I want a kaboomjs dev environment in the browser with a tilemap editor and this extension here is the missing link between the two.

Yeh, Tiled already serves this purpose for me. It's pretty good. I have a long history with Tiled. Before Kaboom even existed, I helped build the current plugin system it has. Before that I used it for years. I am a big fan. It's really well-made software. As a related sidenote, have you seen [Textyle](https://textyle.app/edit/#/tiles)? it's works on the web, and is open-source & is really nice to use. Maybe you could go add features to that (looks liek they have a pretty list of cool TODOs.)

> it is getting closer

Not really sure what more you expect the guy who made a plugin that loads Tiled maps in kaboom, to do. It fulfills the original need I had, and I happily accept PRs for features I agree would be cool (like animation and shared tilesets.) There are a few issues to track features I care about, feel free to tackle any of them. I will merge any code that looks ok, and try to help you through it, if it's not ok.

> The demo doesn't really have the player interact with things in the tilemap though - for example the waterfall.

It's interacting with the other "enemy" player-sprites (so I only need a single sprite image and don't need to go lookup loading a sub-region of an image.) I think you need to use your imagination a bit, and extrapolate what you want from the demos, and try things out on your own, like write some actual code. It's your usecase, not mine. All the pieces are there. If it were me, needing this in my game, I would write a utility function that takes an object-layer and renders it how I want. I would probly also add stuff to sprites based on object-properties (as I do in the sample above.) For over/under water, you should take the water that is not the part hitting ground and put them on a "definitely over player" layer, then put the z-fighting part (the tiles that touch the ground) on a layer where you do Y-sort, like I do in my demo above.

> Rpg maker mv didn't seem to try to address this while using pixi tilemap, so I was a bit curious how another developer would do it

I don't follow. My above code illustrates how to do that.

There is [a great kaboom video series](https://www.youtube.com/watch?v=XX93O4ZVUZI) where an awesome dev makes a zelda clone. She uses all the same ideas here, but much more complete code, and polished up things like movement and goes over these concepts in depth. She is using manual text maps (instead of converting Tiled maps into text, like this plugin does) but otherwise all the same stuff. They are great videos, and working through them may resolve any of your "a tiled map loader plugin maybe should do X", most of which I will probly disagree with (kaboom itself can do a lot of that stuff, and I am not trying to implement every possible usecase, just load Tiled maps, which this plugin does.)

**Comment by blurymind** [comment]
> Yeh, Tiled already serves this purpose for me. It's pretty good. I have a long history with Tiled. Before Kaboom even existed, I helped build the current plugin system it has. Before that I used it for years. I am a big fan. It's really well-made software. As a related sidenote, have you seen [Textyle](https://textyle.app/edit/#/tiles)? it's works on the web, and is open-source & is really nice to use. Maybe you could go add features to that (looks like they have a pretty nice list of cool TODOs.)

I have yeah. Textyle is not purely javascript though and can not be used as an npm module. That kindof put me off it

I dont care about interaction as much as correct rendering of the sprite when it is behind a tile. That is sort of the thing I wonder this extension can solve without being changed

> There is [a great kaboom video series](https://www.youtube.com/watch?v=XX93O4ZVUZI) where an awesome dev makes a zelda clone...

yeah I know. I love her videos :) she got me into kaboomjs

**Comment by konsumer** [comment]
> A solution is to as you say have two tilemaps - above and bellow player. 

Well, that is one solution, that I personally would use for a specific "zelda type game", but that is my point, I think it's very game/developer specific. I am working on [a side-scroller like super-mario](https://github.com/notnullgames/bernieworld) that would have very different solutions to similar issues. I am finding all kinds of things specific to that, for example, it has really long levels that seem to render wrong, which could be a problem with kaboom or my plugin. **Update**: resolved that. I had height/width inverted. surprised I didn't notice before.

With your particular example, I think there are several different ways to solve the "I want the waterfall to appear over the player, and the ground to be over/under correctly". I feel like the solution I outlined would fix it in that case, in a pretty simple way.

I want to keep this library actually focused on what it's for, and not get off into a lot of edge-cases. I think kaboom itself can do a lot to configure how it all fits together, and just using the map-object to do "special things" that don't fit neatly into "turn a tiled map into a kaboom level". I think the other stuff is a matter of documentation (make actual examples that use the different techniques) into things that should be built into the library. Get to know how to do what you are asking for form the tiled-importer, and you can just do that pretty easily with what it outputs, yourself.

> yeah I know. I love her videos :) she got me into kaboomjs

So good. Like informative, explained well, interesting projects. All together, pretty awesome.


Sidenote: not sure why you changed the title. As I said, you can "load specific layers", that is in fact how this plugin works, and I outlined several ways to do this, above. The example code does it in a loop, because it just shows how easy it is to "render all the layers" but it's a trivial matter of basic javascirpt to load them some other way, and it just uses regular kaboom stuff.

**Comment by konsumer** [comment]
Maybe we need a more concrete example. Why don't you make a demo that has the problems you think my library should solve for you, and I can show you specifically how to solve them, as needed, in an actual demo that will benefit other people?

Keep in mind, I disagree that a general ysorting solution for every game/style is something that is needed in every game that uses tiles, and different style games will have different solutions. I feel similarly about a lot of things, so it will be lots of "use the data in `mabObj` and basic kaboom stuff to do that"
